File: 左程云算法班(牛客网)/1.排序/7.3 KMP.java
public static KMP{
    public static int getIndexOf(String s, String m){
        if (s == null || m == null || s.length() < 1 || m.length() < 1){
            return -1;
        }
        char[] str1 = s.toCharArray();
        char[] str2 = m.toCharArray();

        int str1i = 0;
        int str2i = 0;
        int[] next = getNextArray(str2);

        while (str1i <  str1.length && str2i < str2.length){
            if(str1[str1i] == str2[str2i]){
                str1i++;
                str2i++;
            } else if(next[str2i] == -1){
                str1i++;
            }else{
                str2i = next[str2i];
            }
        }
        return str2i == str2.length ? str1i - str2i : -1;
    }


    public static int[] getNextArray(char[] str2){
        if (str2.length == 1){
            return new int[] { -1 };
        }

        int[] next = new int[str2.length];
        next[0] = -1;
        next[1] = 0;
        int pos = 2;
        int cn = 0;
        while (pos < next.length){
            if (str2[pos - 1] == str2[cn]){
                next[pos++] == ++cn;
            }else if(cn > 0){
                cn = next(cn);
            }else{
                next[pos++] = 0;
            }
        }
        return next;

    }

}

File: 左程云算法班(牛客网)/1.排序/7.3 KMP.py
def KMP(str1, str2):
    m = len(str2)
    n = len(str1)
    if str1 == "" or str2 == "" or len(str1) < 1 or len(str2) < 1:
        return
    i1 = 0
    i2 = 0
    nextArray = getNextArray(str2)
    while (i1 < n) and (i2 < m):
        if str2[i2] == str1[i1]:
            i2 += 1
            i1 += 1
        elif nextArray[i2] == -1:
            i1 += 1
        else:
            i2 = nextArray[i2]
    if i2 == m:
        return i1 - i2
    else:
        return -1


def getNextArray(str2):
    m = len(str2)
    nextArray = [0] * m
    nextArray[0] = -1

    if m == 1:
        return nextArray
    nextArray[1] = 0
    pos = 2
    index = 0
    while pos < m:
        if str2[pos - 1] == str2[index]:
            index += 1
            nextArray[pos] = index
            pos += 1
        elif index > 0:
            index = nextArray[index]
        else:
            pos += 1
    return nextArray


# print(KMP("abcabcdabcde", "abcde"))

def func1(a, b):
    return a + b

map = {"abc": func1}


def work(type, a, b):
    return map[type](a, b)


from unittest import mock
def test_work():
    func1 = mock.Mock(side_effect=[1, 1])
    assert work("abc", 1, 1) == 1



test_work()


